l3build.lua

```diff
@@ -24,50 +24,207 @@ for those people who are interested.
+  ---@return string base
+  local function to_dir_base(path)
+    local dir, base = match(path, "(.*/)(.*)")
+    if not dir then dir, base = "./", path end
+    return dir, base
+  end
+
+  ---Central function to allow launching l3build from a subdirectory
+  ---of a local repository.
+  ---Find `base` in `dir` or one of its parents, returns the container.
+  ---We do not assume that `dir` contains no ".." component
+  ---such that we cannot use `to_dir_base`.
+  ---Instead we append `/..` and rely on lua to do the job.
+  ---The max number of trials is the number of components
+  ---of the absolute path of `dir`, which majorated in the for loop
+  ---Intermediate directories must exist.
+  ---@param dir string must end with '/'
+  ---@param base string relative file or directory name
+  ---@return string|nil dir ends with '/' when non nil
+  local function container(dir, base)
+    for _ in gmatch(dir .. lfs.currentdir(), "[^/]+") do
+      local p = dir .. base
+      if os.rename(p, p) then return dir end -- true iff file or dir at the given path
+      -- synonym of previous line:
+      -- if package.searchpath("?", p, "", "") then return dir end
+      dir = dir .. "../"
+    end
+  end
+
+  local cmd_path = arg[0]
+  local cmd_dir, cmd_base = to_dir_base(cmd_path)
+
+  is_main = cmd_base == "l3build" or cmd_base == "l3build.lua"
+  assert(is_main == not not (match(arg[0], "l3build$") or match(arg[0], "l3build%.lua$")))
+  -- l3b_dir:
+  if cmd_base == "l3build.lua" then -- `texlua foo/bar/l3build.lua ...`
+    launch_dir = cmd_dir
+  elseif cmd_base == "l3build" then
+    launch_dir = kpse_dir
+  else
+    launch_dir = container('./', "l3build.lua") or kpse_dir
+  end
+
+  -- work_dir:
+  if cmd_base == "build.lua" then
+    work_dir = cmd_dir
+  else
+    work_dir = container(cmd_dir, "build.lua") or container(start_dir, "build.lua")
+    if not work_dir then
+      print(arg[0])
+      print("  start:  ".. start_dir)
+      -- print("  work:   ".. work_dir)
+      print("  kpse:   ".. kpse_dir)
+      print("  launch: ".. launch_dir)
+      local dir, base = start_dir, "build.lua"
+      for _ in gmatch(dir .. lfs.currentdir(), "[^/]+") do
+        local p = dir .. base
+        print(p)
+        if os.rename(p, p) then return dir end -- true iff file or dir at the given path
+        dir = dir .. "../"
+      end
+    end
+    assert(work_dir, 'Error: Cannot find configuration file "build.lua"')
+  end
+
+  ---Register the given pakage.
+  ---Lua's require function return either true or a table.
+  ---Here we always return a table.
+  ---@param pkg table|boolean
+  ---@param pkg_name string key in `package.loaded`
+  ---@param name string display name
+  ---@param path string
+  local function register(pkg, pkg_name, name, path)
+    if type(pkg) ~= "table" then pkg = {} end
+    package.loaded[pkg_name] = pkg
+    pkg.PACKAGE = pkg_name
+    pkg.NAME = name
+    pkg.PATH = path
+    return pkg
+  end
```

Comment:
This is where objects are registered in the package.loaded table.
For standard require, modules are both boolean or table values.
Here modules are table values, each with informations on itself.